### ISSUE TITLE
Applying Magnet Alignment Algorithm to newly opened Component Browser

### ISSUE THREAD

**Issue by wdanilo** [enso-org/enso#4273]
This task is automatically imported from the old Task Issue Board and it was originally created by Adam Obuchowicz.
Original issue is [here](https://www.pivotaltracker.com/story/show/181076594).
<hr />

## Why
**As** Enso IDE user
**I want** to have new nodes' positions adjusted in situations described in the [design doc](https://github.com/enso-org/design/blob/main/epics/component-browser/design.md#placement-of-newly-opened-component-browser)
**So that** my graph will be nicely laid out.

## Notes

To summarize, we shall apply magnet algorithm in the following cases:
1. When creating node with (+) button without nodes selected
2. When creating node with "Mouse pointer dictated placement." not under the source node
3. When the node is pushed left due to lack of space - only horizontally

## Acceptance Criteria

```gherkin
Scenario: basic cases
GIVEN the following graph:

(====)                            (====)           x
                      or

  x      (====)                                   (====)


WHEN I open component browser near (but not exactly at) the place marked as "x":
a) by pressing tab (without nodes selected)
b) by dragging out connection 
b) by moving the screen accordingly and clicking (+) button
THEN the Component Browser input shall be aligned to both visible nodes.

Scenario: Alignment after opening with (+) button and pushing down to the free space
GIVEN the following graph:
(x===)
   (====)
                      (====)
where the distance between middle and bottom node is a bit bigger from the standard gap between nodes and the scene is centered near (but not exactly at) "x"
WHEN I open component browser by pressing (+) button, 
THEN the Component Browser input shall be aligned horizontally to the top node and vertically to the rightmost node.

Scenario: Alignment after pushing left to free space
GIVEN the following graph
(====) [====]                     [====]
        (====)        or           (====)
                           (====)
where [====] node is selected and the gap between it and the leftmost node is the standard gap between nodes
WHEN I open component browser with tab
THEN the component browser input is aligned to the leftmost node horizontally.
```

## Implementation Notes

1. The Magnet Algorithm is currently available through the `GraphEditor::check_grid_magnet` function.
2. The placement of a newly opened Component Browser is currently done in the `GraphEditorModelWithNetwork::create_node` function. (As of PR 3301, specifically in the `new_node_position::new_node_position` function.)
3. When a node is "pushed to the left due to lack of space" relative to a reference node, its horizontal position becomes different than the horizontal position of the reference node. (As of PR 3301, "pushing to the left due to lack of space" is implemented using the `new_node_position::on_ray` function when called with a `direction` argument of `Vector2(-1.0, 0.0)`.)
4. Applying the Magnet Algorithm to a newly created node in "mouse dictated placement" scenarios needs to be done *iff*:
    - the node is "pushed to the left due to lack of space" (the Magnet Algorithm should be applied only to the horizontal coordinate of the Node position - see point (3) above);
    - or, the node is placed exactly at the mouse position (the Magnet Algorithm should be applied to both coordinates of the Node position).

## Blockers:
[#181076066](https://www.pivotaltracker.com/story/show/#181076066) resolved

## Comments:
A few questions for acceptance criteria:
1. Scenario 2. `distance ... differs slightly` - what means the word `slightly` here? Should it be smaller or larger? 
2. Scenario 3. Is input aligned vertically or horizontally to the leftmost node?  (Ilya Bogdanov - Jan 31, 2022)
<hr />
1. Let it be bigger
2. Horizontally. 
I've updated description (Adam Obuchowicz - Jan 31, 2022)
<hr />
**Mateusz Czapliński** reports a new **STANDUP** for the last Friday (2022-03-18):

**Progress:** Started working on Tech Design. Read the Design Doc & started reading the code related to the magnet/snapping algorithm in Graph Editor. Started writing a draft of Implementation Details. It should be finished by 2022-03-21. (Enso Bot - Mar 21, 2022)
<hr />
**Technical Design** review set to **pass** (Mateusz Czapliński - Mar 21, 2022)
<hr />
**Mateusz Czapliński** reports a new **STANDUP** for today (2022-03-21):

**Progress:** Completed the Tech Design work (added Implementation Notes). Requested a review of the result by AO. It should be finished by 2022-03-21. (Enso Bot - Mar 21, 2022)
<hr />
**Ilya Bogdanov** reports a new **🔴 DELAY** for yesterday (2022-03-22):

**Summary:** There is **4 days** delay in implementation of the [Applying Magnet Alignment Algorithm to newly opened Component Browser (#181076594) task](https://www.pivotaltracker.com/story/show/181076594).
It **will cause 0 days delay** for the delivery of this weekly plan.

**Delay Cause:** the previous standup was reported while making a tech design by another person (Enso Bot - Mar 23, 2022)
<hr />
**Ilya Bogdanov** reports a new **STANDUP** for yesterday (2022-03-22):

**Progress:** started draft implementation, I would need a rebase later to integrate changes from one of not-yet-merged PRs. It should be finished by 2022-03-25.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/181076594). continue working on implementation (Enso Bot - Mar 23, 2022)
<hr />
**Ilya Bogdanov** reports a new **STANDUP** for yesterday (2022-03-23):

**Progress:** rebased onto PR by MC, a lot of manual testing It should be finished by 2022-03-25.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/181076594). polishing the implementation and preparing a PR (Enso Bot - Mar 24, 2022)
<hr />
**Ilya Bogdanov** reports a new **STANDUP** for yesterday (2022-03-24):

**Progress:** mostly playing with different constants for alignment area size & magnet alignment threshold. The result looks okay, but more tweaking might be required. It should be finished by 2022-03-25.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/181076594). polishing the code, also need to take a look at several recent regressions that we've found previously, but didn't investigate yet (Enso Bot - Mar 25, 2022)
<hr />
**Ilya Bogdanov** reports a new **🔴 BLOCKER** for the last Friday (2022-03-25):

**Summary:** The [Applying Magnet Alignment Algorithm to newly opened Component Browser (#181076594) task](https://www.pivotaltracker.com/story/show/181076594) is blocked by the [#181076066 task](https://www.pivotaltracker.com/story/show/181076066).

The blocker size is hard to estimate. The source code heavily depends on not-yet-merged PR 3301

**Possible solutions:** This blocker relation between task should be set while planning (Enso Bot - Mar 28, 2022)
<hr />
**Ilya Bogdanov** reports a new **🔴 DELAY** for the last Friday (2022-03-25):

**Summary:** There is **3 days** delay in implementation of the [Applying Magnet Alignment Algorithm to newly opened Component Browser (#181076594) task](https://www.pivotaltracker.com/story/show/181076594).
It **will cause 1 day delay** for the delivery of this weekly plan.

this delay duplicates blocker

**Delay Cause:** implementation depends on another task, the delay might be extended (Enso Bot - Mar 28, 2022)
<hr />
**Ilya Bogdanov** reports a new **STANDUP** for the last Friday (2022-03-25):

**Progress:** thinking and experimenting about a bug: alignment algorithm places nodes on top of one another in some cases. Found a good enough solution. It should be finished by 2022-03-28.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/181076594). creating a PR, acceptance for task #180887079 (Enso Bot - Mar 28, 2022)
<hr />
**Ilya Bogdanov** reports a new **STANDUP** for yesterday (2022-03-28):

**Progress:** created a PR, recorded a screecast. The PR depends on task #180887079 It should be finished by 2022-03-28. (Enso Bot - Mar 29, 2022)
<hr />
**Ilya Bogdanov** reports a new **🔴 DELAY** for yesterday (2022-03-31):

**Summary:** There is **7 days** delay in implementation of the [Applying Magnet Alignment Algorithm to newly opened Component Browser (#181076594) task](https://www.pivotaltracker.com/story/show/181076594).
It **will cause 1 day delay** for the delivery of this weekly plan.

**Delay Cause:** The task was blocked by another one, also QA reviewer may start working on it only on Monday (Enso Bot - Apr 1, 2022)
<hr />
**Ilya Bogdanov** reports a new **STANDUP** for yesterday (2022-03-31):

**Progress:** rebased onto develop, opened a PR for review. It should be finished by 2022-04-04. (Enso Bot - Apr 1, 2022)
<hr />
**Acceptance/QA** review set to **pass** (Mateusz Czapliński - Apr 4, 2022)
<hr />
**Mateusz Czapliński** reports a new **STANDUP** for yesterday (2022-04-04):

**Progress:** Did Acceptance/QA review manually (passed) and wrote automated tests for the feature. It should be finished by 2022-04-04. (Enso Bot - Apr 5, 2022)
<hr />
